bundles/org.openhab.core.audio/src/main/java/org/openhab/core/audio/internal/AudioServlet.java

```diff
@@ -159,58 +164,156 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
+            StreamServed streamServed = servedStreams.remove(streamId);
+            if (streamServed != null) {
+                tryClose(streamServed.audioStream);
+                // we can notify the caller of the stream consumption
+                streamServed.callback.run();
+                logger.debug("Removed timed out stream {}", streamId);
+            }
         });
+
+        // Because the callback should be executed as soon as possible,
+        // we cannot wait for the next doGet to perform a clean. So we have to schedule a periodic cleaner.
+        ScheduledFuture<?> periodicCleanerLocal = periodicCleaner;
+        if (!servedStreams.isEmpty()) {
+            if (periodicCleanerLocal == null || periodicCleanerLocal.isDone()) {
+                // reschedule a clean
+                periodicCleaner = threadPool.scheduleWithFixedDelay(this::removeTimedOutStreams, 5, 5,
+                        TimeUnit.SECONDS);
+            }
+        } else if (periodicCleanerLocal != null) { // no more stream to serve, shut the periodic cleaning thread:
+            periodicCleanerLocal.cancel(true);
+            periodicCleaner = null;
+        }
     }
 
     @Override
     public String serve(AudioStream stream) {
         String streamId = UUID.randomUUID().toString();
-        oneTimeStreams.put(streamId, stream);
+        // Because we cannot wait indefinitely before executing the callback,
+        // even if this is a one time stream, we set a timeout just in case.
+        long timeOut = System.nanoTime() + TimeUnit.SECONDS.toNanos(10);
+        StreamServed streamToServe = new StreamServed(streamId, stream, new AtomicInteger(), timeOut, false, doNothing);
+        servedStreams.put(streamId, streamToServe);
+
+        // try to clean, or a least launch the periodic cleanse:
+        removeTimedOutStreams();
+
         return getRelativeURL(streamId);
     }
 
     @Override
-    public String serve(FixedLengthAudioStream stream, int seconds) {
+    public String serve(AudioStream stream, int seconds) {
+        return serve(stream, seconds, doNothing);
+    }
+
+    @Override
+    public String serve(AudioStream stream, int seconds, Runnable callBack) {
+
         String streamId = UUID.randomUUID().toString();
-        multiTimeStreams.put(streamId, stream);
-        streamTimeouts.put(streamId, System.nanoTime() + TimeUnit.SECONDS.toNanos(seconds));
+        ClonableAudioStream clonableAudioStream = null;
+        if (stream instanceof ClonableAudioStream clonableAudioStreamDetected) {
+            clonableAudioStream = clonableAudioStreamDetected;
+        } else { // we prefer a ClonableAudioStream, but we can try to make one
+            try {
+                clonableAudioStream = createClonableInputStream(stream, streamId);
+            } catch (AudioException | IOException e) {
+                logger.warn("Cannot precache the audio stream to serve it", e);
+                return getRelativeURL(streamId);
```

Comment:
You are storing neither a streamToServe nor a timeout in that case. Isn't it a problem ?

Reply:
The `serve` method doesn't have any Exception in its signature that I can use, and I tried to avoid any breaking changes.
I'm not satisfied with that either. The issue is then delayed to when the device will try to get the sound. It will get a 404, and it's far from ideal, indeed.
Should I launch a RuntimeException ? But it seems a little bit overkill.

I've added a IOException to the new serve method, at least this one will be OK.
But not to the legacy one... WDYT ? Should we make a Breaking Change ?
Or do you have another idea ?